### PR TITLE
Backport Bitswap stability fix

### PR DESCRIPTION
**Pull request by peterargue**
Backports: https://github.com/onflow/flow-go/pull/2808

### CODE REVIEW

File: cmd/access/node_builder/access_node_builder.go [hunk 1]
```diff
@@ -923,55 +934,79 @@ func (builder *FlowAccessNodeBuilder) Build() (cmd.Node, error) {
...
+				return nil, err
+			}
+
+			builder.AccessNodeConfig.PublicNetworkConfig.Network = net
+
+			node.Logger.Info().Msgf("network will run on address: %s", builder.PublicNetworkConfig.BindAddress)
+			return net, nil
+		}).
+		Component("peer manager", func(node *cmd.NodeConfig) (module.ReadyDoneAware, error) {
+			// TODO: make this component start after its dependencies are ready. This will require a dependable type of component
```
Comment (peterargue) [inline cmd/access/node_builder/access_node_builder.go:985]: this is for the public network, and isn't necessary until we enable bitswap for observers

**Review by yhassanzadeh13 — APPROVED** [review]

File: network/p2p/middleware.go [hunk 2]
```diff
@@ -179,7 +167,9 @@ func NewMiddleware(
 			ready()
 
 			<-ctx.Done()
-			mw.stop()
+
+			// wait for the readConnection and readSubscription routines to stop
+			mw.wg.Wait()
```
Comment (zhangchiqing) [inline network/p2p/middleware.go:172]: Could we add some log how long this wait will take?

I'm a bit concerned this Wait will take too long.

File: cmd/scaffold.go [hunk 3]
```diff
@@ -263,44 +266,65 @@ func (fnb *FlowNodeBuilder) EnqueueNetworkInit() {
...
+
+	fnb.Module("middleware dependency", func(node *NodeConfig) error {
+		fnb.middlewareDependable = module.NewProxiedReadyDoneAware()
+		fnb.PeerManagerDependencies = append(fnb.PeerManagerDependencies, fnb.middlewareDependable)
+		return nil
+	})
+
+	fnb.DependableComponent("peer manager", func(node *NodeConfig) (module.ReadyDoneAware, error) {
+		return fnb.LibP2PNode.PeerManagerComponent(), nil
+	}, fnb.PeerManagerDependencies)
```
Comment (zhangchiqing) [inline cmd/scaffold.go:310]:
```suggestion
	// peer manager won't be created until all PeerManagerDependencies are ready.
	fnb.DependableComponent("peer manager", func(node *NodeConfig) (module.ReadyDoneAware, error) {
		return fnb.LibP2PNode.PeerManagerComponent(), nil
	}, fnb.PeerManagerDependencies)
```

File: cmd/access/node_builder/access_node_builder.go [hunk 4]
```diff
@@ -428,13 +429,22 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionDataRequester() *FlowAccessN
...
+		}).
+		Component("execution data downloader", func(node *cmd.NodeConfig) (module.ReadyDoneAware, error) {
 			var err error
 			bs, err = node.Network.RegisterBlobService(channels.ExecutionDataService, ds)
 			if err != nil {
 				return nil, fmt.Errorf("could not register blob service: %w", err)
 			}
 
+			// add blobservice into ReadyDoneAware dependency passed to peer manager
+			// this configures peer manager to wait for the blobservice to be ready before starting
```
Comment (zhangchiqing) [inline cmd/access/node_builder/access_node_builder.go:445]:
```suggestion
			// this starts the blob service and configures peer manager to wait for the blobservice 
			// to be ready before starting
```

**Review by zhangchiqing — COMMENTED** [review]
Looks good. Just had one question about the Wait

File: network/p2p/middleware.go (hunk 2 above)
Comment (peterargue) [reply network/p2p/middleware.go:172]: sure. this was taken directly from the previous version though, so there's no change in behavior

Comment (zhangchiqing) [reply network/p2p/middleware.go:172]: Can you point out where is the code from the pervious version?

Github shows the change as changing from `mv.stop()` to `mv.wg.Wait()`.  So I thought the logic here is changed, the mv is now stopped elsewhere, and here, we are simply waiting for the result of the `stop`

Comment (peterargue) [reply network/p2p/middleware.go:173]: the original call was here:
https://github.com/onflow/flow-go/pull/3142/files#diff-95120943dbf1b8334da6c35dc229d380cea1f66bb6751a652aab627b9082ffa8L352

This PR converts the peer manager and libp2p node to components, so middleware no longer needs to manage their shutdown. `stop()` became just that single line so I removed the method

**Review by zhangchiqing — APPROVED** [review]
Looks good

**Pull request merged**